### PR TITLE
[LinalgExt] Generalize attention tiling interface implementation

### PR DESCRIPTION
**Pull request by Groverkss**
This patch generalizes tiling implementation for AttentionOp. Before, only the batch and M dimension of attention could be tiled. This patch instead, allows tiling of N dimension as well as allows transposition based on indexing maps (hardcoded for now).

Tiling on dimension N is disabled in CPU backend for now, because TileAndDecomposeAttention pass is hardecoded with dimensions. This will be fixed once we implement reduction tiling interface for it (after https://github.com/llvm/llvm-project/pull/92624)

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/LinalgExt/Utils/IndexingUtils.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+public:
+  AttentionOpDetail(AttentionOp attnOp);
+
+  int64_t getDomainRank() const { return shape.size(); }
+  ArrayRef<int64_t> getDimSizes() const { return shape; }
+  ArrayRef<int64_t> getBatchDims() const { return batch; }
+  ArrayRef<int64_t> getMDims() const { return m; }
+  ArrayRef<int64_t> getK1Dims() const { return k1; }
+  ArrayRef<int64_t> getK2Dims() const { return k2; }
+  ArrayRef<int64_t> getNDims() const { return n; }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Utils/IndexingUtils.h:49]: Could we have more than one batch/m/k1/k2/n dimensions? I thought they all have a single element in each ArrayRef/SmallVector?

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp [hunk 2]
```diff
@@ -1605,69 +1606,110 @@ LogicalResult WinogradOutputTransformOp::getResultTilePosition(
...
+  for (auto dim : llvm::seq<int64_t>(0, opInfo.getDomainRank())) {
     loopBounds[dim].offset = zero;
-    loopBounds[dim].size = getDimValue(builder, loc, source, dim);
     loopBounds[dim].stride = one;
   }
+
+  SmallVector<bool> dimsFound(opInfo.getDomainRank(), false);
+  auto fillSizes = [&](Value val, AffineMap indexingMap) {
+    for (auto [idx, dimExpr] : llvm::enumerate(indexingMap.getResults())) {
+      assert(isa<AffineDimExpr>(dimExpr));
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp:1627]: nit: the assertion is not needed because `cast` already asserts it for you.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp [hunk 3]
```diff
@@ -1605,69 +1606,110 @@ LogicalResult WinogradOutputTransformOp::getResultTilePosition(
...
+      }
+      dimsFound[pos] = true;
+      loopBounds[pos].size = getDimValue(builder, loc, val, idx);
+    }
+  };
+
+  // Sizes can be found from Q, K, V alone.
+  fillSizes(getQuery(), indexingMaps[0]);
+  fillSizes(getKey(), indexingMaps[1]);
+  fillSizes(getValue(), indexingMaps[2]);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: Instead of inferring them from indexingMap (and do we really need indexing maps?), can we make them fixed? Different from linalg op, LinalgExt ops have more strict definition. I'm worried that we are making logics hard to follow while it could just be simple.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 4]
```diff
@@ -1689,14 +1690,71 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
...
+  for (auto i : llvm::seq<unsigned>(0, vecTileSizes.size())) {
+    // Do not tile reduction dimensions.
+    if (vecTileSizes[i] == 0) {
+      continue;
+    }
+    auto tileSize = distTileSizes[i] ? distTileSizes[i] : shapes[i];
+    // TODO: Ask Hanhan what the native size problem here is and change the
+    // "tileSize" arguement to getMaxVectorTileSize accordingly.
+    vecTileSizes[i] = getMaxVectorTileSize(
+        /*numElem=*/tileSize, vectorSize, vectorSize);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: native_vector_size usually means the number of bits of a register on the target cpu. We need to change the `tileSize` to be a factor of problem size because we want to vectorize it without masking. It is a workaround to bypass scf.for bufferization issue.

**Review by hanhanW — COMMENTED** [review]
Some high-level questions:

1. Why do we need indexing_maps for the attention op?
2. Could we have more than 1 dimension for batch/m/n/k1/k2?
3. It looks like tiling on reduction dims in `getTiledImplementation()` is hard. Do you plan to implement [PartialReductionOpInterface](https://github.com/llvm/llvm-project/blob/ec36145f58d2cf93d86bc4e3be617ad7d7d8ace7/mlir/include/mlir/Interfaces/TilingInterface.td#L159) for the attention op?

**Comment by Groverkss** [comment]
> Some high-level questions:
> 
> 1. Why do we need indexing_maps for the attention op?
> 2. Could we have more than 1 dimension for batch/m/n/k1/k2?
> 3. It looks like tiling on reduction dims in `getTiledImplementation()` is hard. Do you plan to implement [PartialReductionOpInterface](https://github.com/llvm/llvm-project/blob/ec36145f58d2cf93d86bc4e3be617ad7d7d8ace7/mlir/include/mlir/Interfaces/TilingInterface.td#L159) for the attention op?

1. The indexing maps are so we can do more fusions with the attention op. One of the fusions that we want to do is fusing transposes with AttentionOp. This PR just hardcodes them, but eventually, they are going to be replaced with indexing maps on the op.

2. Once I add indexing maps, it should be possible to have more than 1 dimensions for batch/m/n/k1/k2. So eventually yes, but not right now. The pytorch SDPA op actually has 2 batch dimensions, so this is it is common to have multiple dimensions for these things.

3. Eventually yes. The plan is to move TileAndDecomposeAttention into Tiling (PartialReductionOpInterface) and Decompose (AggregateOpInterface). This needs some changes upstream, but eventually, this will happen.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Utils/IndexingUtils.h (hunk 1 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Utils/IndexingUtils.h:49]: For now no. But this PR is in preparation for generalizing this, so eventually we will.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp (hunk 3 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: Eventually, we want to have indexing_maps on linalg_ext op so we can do fusions with that op. A simple example is transpose fused with attention op.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 4 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Ah ok makes sense.

**Review by hanhanW — COMMENTED** [review]
Thanks for filling in me the context, it looks more reasonable to me now. The idea sounds okay to me. Is the PR ready for review? I'm seeing some code being commented out in `AttentionOp::verify`

**Comment by Groverkss** [comment]
> Thanks for filling in me the context, it looks more reasonable to me now. The idea sounds okay to me. Is the PR ready for review? I'm seeing some code being commented out in `AttentionOp::verify`

Fixed. It should be ready to review now

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp [hunk 5]
```diff
@@ -1608,23 +1609,55 @@ LogicalResult WinogradOutputTransformOp::getResultTilePosition(
...
+  AttentionOpDetail opInfo = maybeOpInfo.value();
+
+  // All dimensions other than k1 and k2 are parallel.
   SmallVector<utils::IteratorType> iteratorTypes(getIterationDomainRank(),
                                                  utils::IteratorType::parallel);
+
+  for (int dim :
+       llvm::concat<const int64_t>(opInfo.getK1Dims(), opInfo.getK2Dims())) {
+    iteratorTypes[dim] = utils::IteratorType::reduction;
+  }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: The type mismatch, perhaps we just use `auto` for `dim`. Because `concat<>` already spells the type.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp [hunk 6]
```diff
@@ -1637,44 +1670,70 @@ AttentionOp::getTiledImplementation(OpBuilder &builder,
...
+           "Indexing map should be a projected permutation");
+    SmallVector<OpFoldResult> outputOffsets;
+    SmallVector<OpFoldResult> outputSizes;
+    SmallVector<OpFoldResult> outputStrides(indexingMap.getNumResults(), one);
+    for (AffineExpr dimExpr : indexingMap.getResults()) {
+      auto dim = cast<AffineDimExpr>(dimExpr);
+      outputOffsets.push_back(offsets[dim.getPosition()]);
+      outputSizes.push_back(sizes[dim.getPosition()]);
+    }
+    return {outputOffsets, outputSizes, outputStrides};
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp:1687]: nit: use `std::make_tuple()`

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp [hunk 7]
```diff
@@ -1637,44 +1670,70 @@ AttentionOp::getTiledImplementation(OpBuilder &builder,
...
+                    SmallVector<OpFoldResult>> {
+    assert(indexingMap.isProjectedPermutation() &&
+           "Indexing map should be a projected permutation");
+    SmallVector<OpFoldResult> outputOffsets;
+    SmallVector<OpFoldResult> outputSizes;
+    SmallVector<OpFoldResult> outputStrides(indexingMap.getNumResults(), one);
+    for (AffineExpr dimExpr : indexingMap.getResults()) {
+      auto dim = cast<AffineDimExpr>(dimExpr);
+      outputOffsets.push_back(offsets[dim.getPosition()]);
+      outputSizes.push_back(sizes[dim.getPosition()]);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: How about we make it `auto pos = cast<AffineDimExpr>(dimExpr).getPosition()`, so we don't need to two `getPosition()`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 8]
```diff
@@ -1703,14 +1704,82 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
...
+  // these dimensions as subgroups can hold much larger register sizes.
+  TileSizesListType tileSizes;
+
+  // Batch, M and N (parallel dimensions) are distributed on workgroups.
+  DistributionHeuristicConfig config;
   SmallVector<int64_t> distTileSizes = getDefaultDistributedLevelTileSizes(
       attnOp, DistributionHeuristicConfig{});
-  int64_t iterationDomainRank = attnOp.getIterationDomainRank();
-  SmallVector<int64_t> vecTileSizes(iterationDomainRank, 1);
+  // TODO (Groverkss): Due to a bug in TileAndDecomposeAttention, N dimension
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Is there a tracking issue? People could get more context about what's happening in the issue.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 9]
```diff
@@ -1703,14 +1704,82 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
...
+    // TODO: Use native tile size here once bufferization is fixed for scf.
+    vecTileSizes[i] = getMaxVectorTileSize(
+        /*numElem=*/tileSize, vectorSize, vectorSize);
+  }
+  // TODO (Groverkss): Due to a bug in TileAndDecomposeAttention, N dimension
+  // cannot be tiled. Remove this once fixed.
+  for (int64_t i : opInfo.getNDims()) {
+    vecTileSizes[i] = 0;
+  }
+  tileSizes.push_back(vecTileSizes);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: How about adjusting the sizes altogether at the end? In this case, we only need one TODO comment; people won't forget to remove one of them when they enable the tiling for N dims. It also follows the convention we have in the file, i.e., declaring `TileSizesListType` at the end of the function.

```cpp
// distTileSizes = ...
// vecTileSizes = ...

//  TODO(#XXXX): Due to a bug in ...
for (int64_t i : opInfo.getNDims()) {
  distTileSizes[i] = 0;
  vecTileSizes[i] = 0;
}

TileSizesListType tileSizes = {distTileSizes, vecTileSizes};
```

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 10]
```diff
@@ -1294,33 +1274,40 @@ LogicalResult AttentionOp::verify() {
...
-  }
-  if (failed(checkShapeRank(op, "key", keyType, rankToCompareWith))) {
-    return failure();
-  }
-  if (failed(checkShapeRank(op, "value", valueType, rankToCompareWith))) {
-    return failure();
-  }
-  if (failed(checkShapeRank(op, "output", outputType, rankToCompareWith))) {
+  // Check shape compatibility based on indexing maps.
+  SmallVector<int64_t> shape(getIterationDomainRank(), -1);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: The concern is having magic number `-1`; expect it to be different from `ShapedType::kDynamic`. They could have the same value. (Actually `ShapedType::kDynamic` was `-1` in the past, and it is internalized to something else today.) What I'd suggest is having a method to compute the shape first, and then you can compare the result of `affineMap.compose(shape)` and the target tensor shape. It avoids the `-1` magic number.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp [hunk 11]
```diff
@@ -1686,18 +1745,31 @@ LogicalResult AttentionOp::getResultTilePosition(
...
+    resultIndexingMap = *getSumMap();
+    break;
+  default:
+    return failure();
   }
-  return failure();
+
+  for (AffineExpr dimExpr : resultIndexingMap.getResults()) {
+    AffineDimExpr dim = cast<AffineDimExpr>(dimExpr);
+    int64_t pos = dim.getPosition();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: How about we collapse them into a single statement? I.e., `int64_t pos = cast<...>(...).getPosition()`

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 9 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: done

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 10 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: I don't think I have a way of computing the shape without using these indexing maps. I added a `foundDims` array so there is no magic number.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp (hunk 5 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: done

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp (hunk 7 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: done

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp (hunk 6 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp:1687]: I'm not sure about this. I prefer this since the tuple type is already explicitly typed out in the function definition. So there is no ambiguity. Can change if there is a preference.

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp (hunk 11 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp]: done

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 8 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: done

**Review by hanhanW — APPROVED** [review]
